flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/linalg/BLAS.java

```diff
@@ -55,24 +78,34 @@ public static void axpy(double a, SparseVector x, DenseVector y) {
 	 * y += a * x .
 	 */
 	public static void axpy(double a, DenseMatrix x, DenseMatrix y) {
-		assert x.m == y.m && x.n == y.n : "Matrix dimension mismatched.";
+		Preconditions.checkArgument(x.m == y.m && x.n == y.n, "Matrix dimension mismatched.");
 		F2J_BLAS.daxpy(x.data.length, a, x.data, 1, y.data, 1);
 	}
 
+	/**
+	 * y[yOffset:yOffset+n] += a * x[xOffset:xOffset+n] .
+	 */
+	public static void axpy(int n, double a, double[] x, int xOffset, double[] y, int yOffset) {
+		F2J_BLAS.daxpy(n, a, x, xOffset, 1, y, yOffset, 1);
```

Comment:
A higher level question:
Maybe it is good to clarify when to use `F2J_BLAS` vs `NATIVE_BLAS`. 

I found a very [interesting read](https://issues.apache.org/jira/browse/SPARK-21305) from SPARK's mailing list and seems like there are some considerations regarding this.

Reply:
Thanks, I have clarified in the inline doc that we should use F2J_BLAS for level-1 routines and NATIVE_BLAS for level-2 and level-3 routines. This is also the practices adopted by SparkML.

The read from SPARK's mailing list you give here indeed shows the pitfalls when using NATIVE_BLAS. It makes clear that the underlying native BLAS libarary should not use multithreading. Fortunately, the default library uses a BLAS version provided by http://www.netlib.org, which is a single-threaded version.